pandas/core/arrays/datetimelike.py

```diff
@@ -798,72 +818,35 @@ def _validate_listlike(
         return value
 
     def _validate_searchsorted_value(self, value):
-        if isinstance(value, str):
-            try:
-                value = self._scalar_from_string(value)
-            except ValueError as err:
-                raise TypeError(
-                    "searchsorted requires compatible dtype or scalar"
-                ) from err
-
-        elif is_valid_nat_for_dtype(value, self.dtype):
-            value = NaT
-
-        elif isinstance(value, self._recognized_scalars):
-            value = self._scalar_type(value)
-
-        elif not is_list_like(value):
-            raise TypeError(f"Unexpected type for 'value': {type(value)}")
-
+        if not is_list_like(value):
+            value = self._validate_scalar(value, "searchsorted", cast_str=True)
         else:
             # TODO: cast_str?  we accept it for scalar
             value = self._validate_listlike(value, "searchsorted")
 
         return self._unbox(value)
 
     def _validate_setitem_value(self, value):
-
         if is_list_like(value):
             value = self._validate_listlike(value, "setitem", cast_str=True)
-
-        elif isinstance(value, self._recognized_scalars):
-            value = self._scalar_type(value)
-        elif is_valid_nat_for_dtype(value, self.dtype):
-            value = NaT
         else:
-            msg = (
-                f"'value' should be a '{self._scalar_type.__name__}', 'NaT', "
-                f"or array of those. Got '{type(value).__name__}' instead."
-            )
-            raise TypeError(msg)
```

Comment:
This message for example was a nice user-friendly one.

Reply:
ok @gfyoung ?